src/storage/src/source/healthcheck.rs

```diff
@@ -0,0 +1,809 @@
+use chrono::NaiveDateTime;
+use std::fmt::Display;
+use std::sync::Arc;
+use timely::progress::{Antichain, Timestamp as _};
+use timely::PartialOrder;
+use tokio::sync::Mutex;
+use tracing::trace;
+
+use mz_persist_client::cache::PersistClientCache;
+use mz_persist_client::read::{Listen, ListenEvent, ReadHandle};
+use mz_persist_client::write::WriteHandle;
+use mz_repr::{Datum, GlobalId, Row, Timestamp};
+
+use crate::source::{CollectionMetadata, NowFn};
+use crate::types::sources::SourceData;
+
+/// The Healthchecker is responsible for tracking the current state
+/// of a Timely worker for a source, as well as updating the relevant
+/// state collection based on it.
+pub struct Healthchecker {
+    /// Name of the source (e.g. kafka-s1)
+    source_name: String,
+    /// Name of the upstream resource, if any (e.g. Kafka topic, file path)
+    upstream_name: Option<String>,
+    /// Internal ID of the source (e.g. s1)
+    source_id: String,
+    /// Type of the source, from SourceConnection::name
+    source_type: &'static str,
+    /// The ID of the Timely worker on which this operator is executing
+    worker_id: usize,
+    /// The total count of Timely workers
+    worker_count: usize,
+    /// Whether this is an active Timely worker or not
+    active: bool,
+    /// Current status of this source
+    current_status: SourceStatus,
+    /// Last observed upper
+    upper: Antichain<Timestamp>,
+    /// Write handle of the Healthchecker persist shard
+    write_handle: WriteHandle<SourceData, (), Timestamp, i64>,
+    /// A listener to tail the Healthchecker shard for new updates
+    listener: Listen<SourceData, (), Timestamp, i64>,
+    /// The function that should be used to get the current time when minting new bindings
+    now: NowFn,
+}
+
+impl Healthchecker {
+    pub async fn new(
+        source_name: String,
+        upstream_name: Option<String>,
+        source_id: GlobalId,
+        source_type: &'static str,
+        worker_id: usize,
+        worker_count: usize,
+        active: bool,
+        persist_clients: &Arc<Mutex<PersistClientCache>>,
+        storage_metadata: &CollectionMetadata,
+        now: NowFn,
+    ) -> anyhow::Result<Self> {
+        trace!("Initializing healthchecker for source {source_name}");
+        let mut persist_clients = persist_clients.lock().await;
+        let persist_client = persist_clients
+            .open(storage_metadata.persist_location.clone())
+            .await
+            .context("error creating persist client for Healthchecker")?;
+        drop(persist_clients);
+
+        let (write_handle, read_handle) = persist_client
+            .open(storage_metadata.status_shard)
+            .await
+            .context("error opening Healthchecker persist shard")?;
+
+        let (since, upper) = (read_handle.since().clone(), write_handle.upper().clone());
+
+        let listener = read_handle
+            .clone()
+            .await
+            .listen(since.clone())
```

Comment:
Because the `snapshot` happens at the `since`, as it can't happen at the `upper`, and doing `upper - 1` felt a bit too much like hard-coding that our timestamps are `u64`s. And AFAIU, we could lose updates if the `listener` was started at `upper` if the snapshot was generated at `upper`.

Reply:
yeah, this is indeed a bit rough. Let's keep it like this, and maybe add what you commented here as a comment, wdyt?

Reply:
Sounds good to me!